File: orchestrator/app/instrumentation.py
"""Prometheus instruments for the orchestrator.

Aggregate view over RequestMetrics: runtime.py calls record_request()
once per finished request, and /metrics (mounted in main.py) renders
the registry on each Prometheus scrape. Nothing is pushed anywhere.

Label rule: bounded values only (stage, tool, model, outcome) — never
user_id / chat_id / request_id, each unique value is a new time series.

Assumes a single uvicorn worker process (as in the Dockerfile). With
--workers N each process would have its own registry and this needs
prometheus_client's multiprocess mode instead.
"""

from collections import Counter as Multiset

from prometheus_client import Counter, Histogram

from app.metrics import RequestMetrics

# LLM-shaped buckets: the generator stage routinely runs 10-60s, so the
# default buckets (max 10s) would lump everything interesting into +Inf.
STAGE_BUCKETS = (0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 20, 30, 60, 120)

CHAT_REQUESTS = Counter(
    "chat_requests_total",
    "Chat requests handled by the runtime.",
    ["outcome"],  # ok | error
)
# Pre-create both series so rate() queries see 0 instead of "no data"
# before the first error ever happens.
for outcome in ("ok", "error"):
    CHAT_REQUESTS.labels(outcome)

STAGE_DURATION = Histogram(
    "chat_stage_duration_seconds",
    "Wall-clock duration of each runtime stage.",
    ["stage"],
    buckets=STAGE_BUCKETS,
)

LLM_TOKENS = Counter(
    "llm_tokens_total",
    "LLM tokens consumed, by kind and model.",
    ["kind", "model"],   # kind = prompt | completion
)

TOOL_CALLS = Counter(
    "tool_calls_total",
    "Tool calls done, by tool and status.",
    ["tool", "status"] # status = ok | error
)

VERIFIER_ISSUES = Counter(
    "verifier_issues_total",
    "Amount of failed verifier checks",
    ["issue"]
)

DOCS_BUCKETS = (0, 1, 2, 5, 10, 20)

DOCS_RETRIEVED = Histogram(
    "docs_retrieved",
    "Number of documents retrieved per request.",
    buckets=DOCS_BUCKETS
)

PROGRAM_REQUESTS = Counter(
    "program_requests_total",
    "Program AI operations handled by the runtime.",
    ["op", "outcome"],  # op = normalize | suggest
)
# Outcome sets differ per op: only normalize can reject ("that's a recipe");
# suggest either works or errors. Pre-created like CHAT_REQUESTS above.
for _op, _outcomes in (("normalize", ("ok", "rejected", "error")), ("suggest", ("ok", "error"))):
    for _outcome in _outcomes:
        PROGRAM_REQUESTS.labels(_op, _outcome)

PROGRAM_SUGGESTIONS = Counter(
    "program_suggestions_total",
    "Suggestions from the suggest pipeline, by validation status.",
    # dropped/(emitted+dropped) is the LLM-quality signal: a rising drop
    # rate means hallucinated target ids or malformed JSONL.
    ["status"],  # emitted | dropped
)
for _status in ("emitted", "dropped"):
    PROGRAM_SUGGESTIONS.labels(_status)


def record_request(metrics: RequestMetrics, outcome: str, kind: str = "chat") -> None:
    if kind == "chat":
        CHAT_REQUESTS.labels(outcome).inc()
    else:  # "program_normalize" | "program_suggest"
        PROGRAM_REQUESTS.labels(kind.removeprefix("program_"), outcome).inc()
    for stage, ms in metrics.latencies_ms.items():
        STAGE_DURATION.labels(stage).observe(ms / 1000)
    
    model = metrics.generator_model or "none"
    if metrics.prompt_tokens:
        LLM_TOKENS.labels("prompt", model).inc(metrics.prompt_tokens)
    if metrics.completion_tokens:
        LLM_TOKENS.labels("completion", model).inc(metrics.completion_tokens)

    used = Multiset(metrics.tools_used)
    errors = Multiset(metrics.tool_errors)

    for tool, total in used.items():
        error_count = errors.get(tool, 0)
        ok_count = total - error_count

        if ok_count:
            TOOL_CALLS.labels(tool, "ok").inc(ok_count)

        if error_count:
            TOOL_CALLS.labels(tool, "error").inc(error_count)
    
    for issue in metrics.verifier_issues:
        VERIFIER_ISSUES.labels(issue).inc()

    DOCS_RETRIEVED.observe(metrics.docs_retrieved)


def record_suggestions(emitted: int, dropped: int) -> None:
    if emitted:
        PROGRAM_SUGGESTIONS.labels("emitted").inc(emitted)
    if dropped:
        PROGRAM_SUGGESTIONS.labels("dropped").inc(dropped)

File: java-orchestrator/src/main/java/com/powerlifting/orchestrator/observability/Instrumentation.java
package com.powerlifting.orchestrator.observability;

import com.powerlifting.orchestrator.chat.runtime.Verifier;
import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.DistributionSummary;
import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Timer;
import java.time.Duration;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.springframework.stereotype.Component;

/**
 * Aggregate metrics for the runtime.
 *
 * <p>{@link RequestMetrics} is the per-request record (logged, streamed to the
 * gateway, the seed for evaluation); this is the fleet-wide view folded from it
 * exactly once, when {@link #recordRequest} is called at the end of a turn. The
 * meters are exported at {@code /actuator/prometheus} by the autoconfigured
 * Prometheus registry.
 *
 * <p>Meter and tag names are chosen so the Prometheus exposition matches the
 * names the dashboards already query ({@code chat_requests_total},
 * {@code chat_stage_duration_seconds}, {@code llm_tokens_total},
 * {@code tool_calls_total}, {@code verifier_issues_total},
 * {@code docs_retrieved}). Micrometer converts the dotted names and appends the
 * {@code _total} / {@code _seconds} suffixes.
 *
 * <p><b>Cardinality rule (hard):</b> tag values must be bounded — stage, tool,
 * model, outcome, issue, kind. Never user id, chat id or request id; each unique
 * value is a new time series.
 */
@Component
public class Instrumentation {

    public static final String OUTCOME_OK = "ok";
    public static final String OUTCOME_ERROR = "error";

    // Same bucket boundaries the dashboards expect. Stage durations routinely
    // reach tens of seconds (the generation stage), so the buckets extend well
    // past Micrometer's sub-second defaults.
    private static final Duration[] STAGE_SLOS = {
            Duration.ofMillis(50), Duration.ofMillis(100), Duration.ofMillis(250),
            Duration.ofMillis(500), Duration.ofSeconds(1), Duration.ofMillis(2500),
            Duration.ofSeconds(5), Duration.ofSeconds(10), Duration.ofSeconds(20),
            Duration.ofSeconds(30), Duration.ofSeconds(60), Duration.ofSeconds(120)};

    // Micrometer forbids an SLO boundary of 0, so a request that retrieved
    // nothing lands in the le=1 bucket rather than a dedicated le=0 one. The
    // _count/_sum series still capture every request.
    private static final double[] DOCS_SLOS = {1, 2, 5, 10, 20};

    // The stages ChatService times, pre-registered so their series exist at zero.
    private static final List<String> KNOWN_STAGES =
            List.of("planner", "retrieval", "generation", "summary");

    // The tools the model can call, pre-registered × {ok, error}.
    private static final List<String> KNOWN_TOOLS =
            List.of("get_lifter_history", "leaderboard_query");

    private final MeterRegistry registry;
    private final Map<String, Timer> stageTimers = new HashMap<>();
    private final DistributionSummary docsRetrieved;

    public Instrumentation(MeterRegistry registry) {
        this.registry = registry;

        // Pre-register the bounded series so rate()/increase() over them read 0
        // rather than "no data" before the first occurrence — the dashboards
        // depend on this, together with their `or vector(0)` guards.
        for (String outcome : List.of(OUTCOME_OK, OUTCOME_ERROR)) {
            requestCounter(outcome);
        }
        for (String stage : KNOWN_STAGES) {
            stageTimers.put(stage, buildStageTimer(stage));
        }
        for (String tool : KNOWN_TOOLS) {
            toolCounter(tool, OUTCOME_OK);
            toolCounter(tool, OUTCOME_ERROR);
        }
        for (String issue : Verifier.ALL_ISSUES) {
            issueCounter(issue);
        }

        this.docsRetrieved = DistributionSummary.builder("docs.retrieved")
                .description("Documents retrieved per request")
                .serviceLevelObjectives(DOCS_SLOS)
                .register(registry);
    }

    /**
     * Folds one finished request into the aggregate meters.
     *
     * @param outcome {@link #OUTCOME_OK} or {@link #OUTCOME_ERROR}; a client
     *                disconnect is {@code ok}, not an error.
     */
    public void recordRequest(RequestMetrics metrics, String outcome) {
        requestCounter(outcome).increment();

        metrics.latenciesMs().forEach((stage, ms) ->
                stageTimers.computeIfAbsent(stage, this::buildStageTimer)
                        .record(Duration.ofNanos(Math.round(ms * 1_000_000))));

        String model = metrics.generatorModel() != null ? metrics.generatorModel() : "none";
        if (metrics.promptTokens() > 0) {
            tokenCounter("prompt", model).increment(metrics.promptTokens());
        }
        if (metrics.completionTokens() > 0) {
            tokenCounter("completion", model).increment(metrics.completionTokens());
        }

        recordToolCalls(metrics);

        for (String issue : metrics.verifierIssues()) {
            issueCounter(issue).increment();
        }

        docsRetrieved.record(metrics.docsRetrieved());
    }

    /**
     * A tool that appears in {@code toolErrors} once counts as one error and
     * (if it also ran successfully) that many oks — matching how a single
     * request can call the same tool several times.
     */
    private void recordToolCalls(RequestMetrics metrics) {
        Map<String, Long> used = countByName(metrics.toolsUsed());
        Map<String, Long> errored = countByName(metrics.toolErrors());

        used.forEach((tool, total) -> {
            long errors = errored.getOrDefault(tool, 0L);
            long ok = total - errors;
            if (ok > 0) {
                toolCounter(tool, OUTCOME_OK).increment(ok);
            }
            if (errors > 0) {
                toolCounter(tool, OUTCOME_ERROR).increment(errors);
            }
        });
    }

    private static Map<String, Long> countByName(List<String> names) {
        Map<String, Long> counts = new LinkedHashMap<>();
        for (String name : names) {
            counts.merge(name, 1L, Long::sum);
        }
        return counts;
    }

    private Counter requestCounter(String outcome) {
        return registry.counter("chat.requests", "outcome", outcome);
    }

    private Counter tokenCounter(String kind, String model) {
        return registry.counter("llm.tokens", "kind", kind, "model", model);
    }

    private Counter toolCounter(String tool, String status) {
        return registry.counter("tool.calls", "tool", tool, "status", status);
    }

    private Counter issueCounter(String issue) {
        return registry.counter("verifier.issues", "issue", issue);
    }

    private Timer buildStageTimer(String stage) {
        return Timer.builder("chat.stage.duration")
                .description("Wall-clock duration of a runtime stage")
                .tag("stage", stage)
                .serviceLevelObjectives(STAGE_SLOS)
                .register(registry);
    }
}
